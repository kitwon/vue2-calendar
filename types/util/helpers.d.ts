import { Slots } from 'vue';
export declare function escapeHTML(str: string): string;
export declare function genFunctionComputed<T = Function>(prop: any, func?: T): import("vue").ComputedRef<T>;
export declare function genIntComputed(props: any): import("vue").ComputedRef<number>;
export declare function convertToUnit(str: string | number | null | undefined, unit?: string): string | undefined;
export declare function getEventName(str: string): string;
/**
  In vue3, all event in jsx use on[Event] with camel-case.
  eg: `ctx.emit('customEvent')`. In jsx `<component onCustomEvent={handler} />`.

  More info
  https://v3.vuejs.org/guide/migration/render-function-api.html#_3-x-syntax-2

 * @export
 * @param {string} str
 * @returns
 */
export declare function getComponentEvent(str: string): string;
export declare function getTriggerEventName(str: string): string;
export declare function isCssColor(color: string): boolean;
export declare function getColor(styleName: string | string[], color: string, data?: {
    style: string;
    class: string;
}): {
    style: any;
    class: any;
};
export declare function getSlot(slots: Slots, name: string, data?: any): import("vue").VNode<import("vue").RendererNode, import("vue").RendererElement, {
    [key: string]: any;
}>[] | null;
