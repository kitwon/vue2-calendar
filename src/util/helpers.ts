import { computed, Slots } from 'vue'

const tagsToReplace = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;'
} as any

export function escapeHTML(str: string): string {
  return str.replace(/[&<>]/g, tag => tagsToReplace[tag] || tag)
}

export function genFunctionComputed<T = Function>(prop: any, func?: T) {
  return computed<T>(() => {
    return typeof prop === 'function' ? prop : func || (() => prop)
  })
}

export function genIntComputed(props: any) {
  return computed(() => parseInt(props))
}

export function convertToUnit(
  str: string | number | null | undefined,
  unit = 'px'
): string | undefined {
  if (str == null || str === '') {
    return undefined
  } else if (typeof str === 'string') {
    return String(str)
  } else {
    return `${Number(str)}${unit}`
  }
}

export function getEventName(str: string) {
  return str[2].toLowerCase() + str.slice(2)
}

/**
  In vue3, all event in jsx use on[Event] with camel-case.
  eg: `ctx.emit('customEvent')`. In jsx `<component onCustomEvent={handler} />`.

  More info
  https://v3.vuejs.org/guide/migration/render-function-api.html#_3-x-syntax-2

 * @export
 * @param {string} str
 * @returns
 */
export function getComponentEvent(str: string) {
  return `on${str[0].toUpperCase()}${str.slice(1)}`
}

export function getTriggerEventName(str: string) {
  const name = str.replace(/^on(.*)/, '$1')
  return `${name[0].toLocaleLowerCase()}${name.slice(1)}`
}

export function isCssColor(color: string) {
  return !!color && !!color.match(/^(#|var\(--|(rgb|hsl)a?\()/)
}

// Need Refactor
export function getColor(
  styleName: string | string[],
  color: string,
  data = { style: '', class: '' }
) {
  const colorObj = {
    style: [data.style] as any,
    class: [data.class] as any
  }
  if (isCssColor(color)) {
    const keys = Array.isArray(styleName) ? styleName : [styleName]
    colorObj.style.push(
      keys.reduce<Record<string, string>>((acc, key) => {
        acc[key] = color
        return acc
      }, {})
    )
  } else {
    colorObj.class.push(color)
  }

  return colorObj
}

export function getSlot(slots: Slots, name: string, data?: any) {
  const slot = slots[name]
  if (typeof slot !== 'undefined') {
    return slot(data)
  }

  return null
}
