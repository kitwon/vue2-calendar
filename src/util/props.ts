import { ExtractPropTypes, PropType } from 'vue'

//Utils
import {
  validateTimestamp,
  parseDate,
  DAYS_IN_WEEK,
  validateTime
} from './timestamp'

// Types
import {
  CalendarEvent,
  CalendarFormatter,
  CalendarTimestamp,
  CalendarEventOverlapMode,
  CalendarEventNameFunction,
  CalendarEventColorFunction,
  CalendarEventCategoryFunction,
  CalendarEventTimedFunction
} from '../types'
import { CalendarEventOverlapModes } from '../modes'

export function validateNumber(input: any): boolean {
  return isFinite(parseInt(input))
}

export function validateWeekdays(input: string | (number | string)[]): boolean {
  if (typeof input === 'string') {
    input = input.split(',')
  }

  if (Array.isArray(input)) {
    const ints = input.map(x => parseInt(x as string))

    if (ints.length > DAYS_IN_WEEK || ints.length === 0) {
      return false
    }

    const visited: Record<number, boolean> = {}
    let wrapped = false

    for (let i = 0; i < ints.length; i++) {
      const x = ints[i]

      if (!isFinite(x) || x < 0 || x >= DAYS_IN_WEEK) {
        return false
      }

      if (i > 0) {
        const d = x - ints[i - 1]
        if (d < 0) {
          if (wrapped) {
            return false
          }
          wrapped = true
        } else if (d === 0) {
          return false
        }
      }

      if (visited[x]) {
        return false
      }
      visited[x] = true
    }

    return true
  }

  return false
}

const props = {
  base: {
    start: {
      type: [String, Number, Date],
      validate: validateTimestamp,
      default: () => parseDate(new Date()).date
    },
    end: {
      type: [String, Number, Date],
      validate: validateTimestamp
    },
    weekdays: {
      type: [Array, String] as PropType<number[] | string>,
      default: () => [0, 1, 2, 3, 4, 5, 6],
      validate: validateWeekdays
    },
    hideHeader: {
      type: Boolean
    },
    shortWeekdays: {
      type: Boolean,
      default: true
    },
    weekdayFormat: {
      type: Function as PropType<CalendarFormatter>,
      default: null
    },
    dayFormat: {
      type: Function as PropType<CalendarFormatter>,
      default: null
    }
  },
  intervals: {
    maxDays: {
      type: Number,
      default: 7
    },
    shortIntervals: {
      type: Boolean,
      default: false
    },
    intervalHeight: {
      type: [Number, String],
      default: 48,
      validate: validateNumber
    },
    intervalWidth: {
      type: [Number, String],
      default: 60,
      validate: validateNumber
    },
    intervalMinutes: {
      type: [Number, String],
      default: 60,
      validate: validateNumber
    },
    firstInterval: {
      type: [Number, String],
      default: 0,
      validate: validateNumber
    },
    firstTime: {
      type: [Number, String, Object],
      validate: validateTime
    },
    intervalCount: {
      type: [Number, String],
      default: 24,
      validate: validateNumber
    },
    intervalFormat: {
      type: Function as PropType<CalendarFormatter>,
      default: null
    },
    intervalStyle: {
      type: Function as PropType<(interval: CalendarTimestamp) => object>,
      default: null
    },
    showIntervalLabel: {
      type: Function as PropType<(interval: CalendarTimestamp) => boolean>,
      default: null
    }
  },
  weeks: {
    localeFirstDayOfYear: {
      type: [String, Number],
      default: 0
    },
    minWeeks: {
      validate: validateNumber,
      default: 1
    },
    shortMonths: {
      type: Boolean,
      default: true
    },
    showMonthOnFirst: {
      type: Boolean,
      default: true
    },
    showWeek: Boolean,
    monthFormat: {
      type: Function as PropType<CalendarFormatter>,
      default: null
    }
  },
  calendar: {
    /**
     * Calendar type
     */
    type: {
      type: String,
      default: 'month'
    },
    modelValue: {
      type: [String, Number, Date] as PropType<string | number | Date>,
      validate: validateTimestamp
    },
    now: {
      type: String,
      validate: validateTimestamp
    },
    locale: {
      type: String,
      default: 'en-US'
    }
  },
  category: {
    categories: {
      type: [Array, String],
      default: ''
    },
    categoryHideDynamic: {
      type: Boolean
    },
    categoryShowAll: {
      type: Boolean
    },
    categoryForInvalid: {
      type: String,
      default: ''
    },
    categoryDays: {
      type: [Number, String],
      default: 1,
      validate: (x: any) => isFinite(parseInt(x)) && parseInt(x) > 0
    }
  },
  events: {
    events: {
      type: Array as PropType<CalendarEvent[]>,
      default: () => []
    },
    eventStart: {
      type: String,
      default: 'start'
    },
    eventEnd: {
      type: String,
      default: 'end'
    },
    eventTimed: {
      type: [String, Function] as PropType<string | CalendarEventTimedFunction>,
      default: 'timed'
    },
    eventCategory: {
      type: [String, Function] as PropType<
        string | CalendarEventCategoryFunction
      >,
      default: 'category'
    },
    eventHeight: {
      type: Number,
      default: 20
    },
    eventColor: {
      type: [String, Function] as PropType<string | CalendarEventColorFunction>,
      default: 'is-text-default'
    },
    eventTextColor: {
      type: [String, Function] as PropType<string | CalendarEventColorFunction>,
      default: 'is-default'
    },
    eventName: {
      type: [String, Function] as PropType<string | CalendarEventNameFunction>,
      default: 'name'
    },
    eventOverlapThreshold: {
      type: [String, Number],
      default: 60
    },
    eventOverlapMode: {
      type: [String, Function] as PropType<
        'stack' | 'column' | CalendarEventOverlapMode
      >,
      default: 'stack',
      validate: (mode: any) =>
        mode in CalendarEventOverlapModes || typeof mode === 'function'
    },
    eventMore: {
      type: Boolean,
      default: true
    },
    eventMoreText: {
      type: String,
      default: '$v-calendar.moreEvents'
    },
    eventRipple: {
      type: [Boolean, Object],
      default: null
    },
    eventMarginBottom: {
      type: Number,
      default: 1
    }
  }
}

export default props

export type EventProps = ExtractPropTypes<typeof props.events>
export type BaseProps = ExtractPropTypes<
  typeof props.base & typeof props.calendar
>
export type IntervalProps = ExtractPropTypes<typeof props.intervals>
export type CategoryProps = ExtractPropTypes<typeof props.category>
export type CalendarProps = EventProps &
  BaseProps &
  IntervalProps &
  CategoryProps
