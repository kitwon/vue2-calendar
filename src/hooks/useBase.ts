import {
  createDayList,
  createNativeLocaleFormatter,
  getTimestampIdentifier,
  getWeekdaySkips,
  getStartOfWeek,
  getEndOfWeek,
  parseTimestamp
} from '../util/timestamp'
import { CalendarFormatter, CalendarTimestamp } from '../types'
import { computed } from 'vue'
import useTimes from './useTimes'
import { BaseProps } from '../util/props'

export default function useBase(props: BaseProps) {
  const { times, updateTimes } = useTimes(props.now)

  const parsedWeekdays = computed(() => {
    return Array.isArray(props.weekdays)
      ? props.weekdays
      : (props.weekdays || '').split(',').map((x: any) => parseInt(x, 10))
  })
  const weekdaySkips = computed(() => {
    return getWeekdaySkips(parsedWeekdays.value)
  })
  const weekdaySkipsReverse = computed(() => {
    const reversed = weekdaySkips.value.slice()
    reversed.reverse()
    return reversed
  })
  const parsedStart = computed(() => {
    return parseTimestamp(props.start, true)
  })
  const parsedEnd = computed(() => {
    const start = parsedStart.value
    const end: CalendarTimestamp = props.end
      ? parseTimestamp(props.end) || start
      : start

    return getTimestampIdentifier(end) < getTimestampIdentifier(start)
      ? start
      : end
  })
  const days = computed(() => {
    return createDayList(
      parsedStart.value,
      parsedEnd.value,
      times.today,
      weekdaySkips.value
    )
  })

  const dayFormatter = computed(() => {
    if (props.dayFormat) {
      return props.dayFormat
    }

    const options = { timeZone: 'UTC', day: 'numeric' }

    return createNativeLocaleFormatter(props.locale, () => options)
  })
  const weekdayFormatter = computed(() => {
    if (props.weekdayFormat) {
      return props.weekdayFormat
    }

    const longOptions = { timeZone: 'UTC', weekday: 'long' }
    const shortOptions = { timeZone: 'UTC', weekday: 'short' }

    return createNativeLocaleFormatter(props.locale, (_tms, short) =>
      short ? shortOptions : longOptions
    )
  })
  const getFormatter = (options: object): CalendarFormatter => {
    return createNativeLocaleFormatter(props.locale, () => options)
  }
  const getRelativeClasses = (
    timestamp: CalendarTimestamp,
    outside = false
  ) => {
    return {
      'v-present': timestamp.present,
      'v-past': timestamp.past,
      'v-future': timestamp.future,
      'v-outside': outside
    }
  }

  return {
    times,
    parsedWeekdays,
    weekdaySkips,
    weekdaySkipsReverse,
    parsedStart,
    parsedEnd,
    days,
    dayFormatter,
    weekdayFormatter,
    updateTimes,
    getFormatter,
    getRelativeClasses,
    getStartOfWeek: (timestamp: CalendarTimestamp) => {
      return getStartOfWeek(timestamp, parsedWeekdays.value, times.today)
    },
    getEndOfWeek: (timestamp: CalendarTimestamp) => {
      return getEndOfWeek(timestamp, parsedWeekdays.value, times.today)
    }
  }
}
