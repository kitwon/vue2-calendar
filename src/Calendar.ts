// Styles
import './Event.scss'

import { Component, computed, defineComponent, h, reactive, ref } from 'vue'
import props from './util/props'

// Types
import { CalendarTimestamp } from './types'

// Hooks
import useEvents from './hooks/useEvents'
import useBase from './hooks/useBase'
import useIntervals from './hooks/useIntervals'

// import resize from './directives/resize'

// Components
import Daily from './Daily'
import Weekly from './Weekly'

import {
  copyTimestamp,
  DAYS_IN_MONTH_MAX,
  DAYS_IN_WEEK,
  DAY_MIN,
  getEndOfMonth,
  getEndOfWeek,
  getStartOfMonth,
  getStartOfWeek,
  nextDay,
  parseTimestamp,
  prevDay,
  relativeDays,
  timestampToDate,
  updateFormatted,
  updateRelative,
  updateWeekday,
  validateTimestamp,
  VTime
} from './util/timestamp'
// import { genIntComputed } from './util/helpers'

interface VCalendarRenderProps {
  start: CalendarTimestamp
  end: CalendarTimestamp
  component: string | Component
  maxDays: number
  weekdays: number[]
  categories: string[]
}

const Calendar = defineComponent({
  name: 'Calendar',
  props: {
    ...props.base,
    ...props.calendar,
    ...props.weeks,
    ...props.intervals,
    ...props.category,
    ...props.events
  },
  setup(props, ctx) {
    const {
      times,
      parsedStart,
      parsedEnd,
      parsedWeekdays,
      updateTimes
    } = useBase(props)
    const { noEvents, getScopeSlots } = useEvents(props, ctx)
    const { timeToY } = useIntervals(props)
    const calendarRef = ref<any>(null)
    const history = reactive({
      lastStart: null as CalendarTimestamp | null,
      lastEnd: null as CalendarTimestamp | null
    })

    const parsedValue = computed(() => {
      return validateTimestamp(props.modelValue)
        ? parseTimestamp(props.modelValue, true)
        : parsedStart.value
    })

    const getCalendarStartOfWeek = (timestamp: CalendarTimestamp) => {
      return getStartOfWeek(timestamp, parsedWeekdays.value, times.today)
    }
    const getCalendarEndOfWeek = (timestamp: CalendarTimestamp) => {
      return getEndOfWeek(timestamp, parsedWeekdays.value, times.today)
    }

    const renderProps = computed(() => {
      const around = parsedValue.value
      let component: any = null
      let maxDays = props.maxDays
      let weekdays = parsedWeekdays.value
      let classes = ''
      let start = around
      let end = around
      switch (props.type) {
        case 'month':
          component = Weekly
          start = getStartOfMonth(around)
          end = getEndOfMonth(around)
          classes = 'v-calendar-month'
          break
        case 'week':
          component = Daily
          start = getCalendarStartOfWeek(around)
          end = getCalendarEndOfWeek(around)
          maxDays = 7
          break
        case 'day':
          component = Daily
          maxDays = 1
          weekdays = [start.weekday]
          break
        case 'custom-weekly':
          component = Weekly
          start = parsedStart.value || around
          end = parsedEnd.value
          break
        case 'custom-daily':
          component = Daily
          start = parsedStart.value || around
          end = parsedEnd.value
          break
        default:
          throw new Error(props.type + ' is not a valid Calendar type')
      }

      return {
        component,
        classes,
        props: { start, end, maxDays, weekdays }
      }
    })

    // ------------------ Calendar methods ---------------------
    const move = (amount = 1) => {
      const moved = copyTimestamp(parsedValue.value)
      const forward = amount > 0
      const mover = forward ? nextDay : prevDay
      const limit = forward ? DAYS_IN_MONTH_MAX : DAY_MIN
      let step = forward ? amount : -amount

      while (--step >= 0) {
        switch (props.type) {
          case 'month':
            moved.day = limit
            mover(moved)
            break
          case 'week':
          case 'custom-weekly':
            relativeDays(moved, mover, DAYS_IN_WEEK)
            break
          case 'day':
            relativeDays(moved, mover, 1)
            break
          case '4day':
            relativeDays(moved, mover, 4)
            break
        }
      }

      updateWeekday(moved)
      updateFormatted(moved)
      updateRelative(moved, times.now)

      if (props.modelValue instanceof Date) {
        ctx.emit('update:modelValue', timestampToDate(moved))
      } else if (typeof props.modelValue === 'number') {
        ctx.emit('update:modelValue', timestampToDate(moved).getTime())
      } else {
        ctx.emit('update:modelValue', moved.date)
      }

      ctx.emit('moved', moved)
    }
    const next = (amount = 1) => {
      move(amount)
    }
    const prev = (amount = 1) => {
      move(-amount)
    }

    const scrollToTime = (time: VTime) => {
      const c = calendarRef.value

      if (c && c.scrollToTime) {
        return c.scrollToTime(time)
      } else {
        return false
      }
    }

    const checkChange = () => {
      const { lastStart, lastEnd } = history
      const { start, end } = renderProps.value.props
      if (
        !lastStart ||
        !lastEnd ||
        start.date !== lastStart.date ||
        end.date !== lastEnd.date
      ) {
        history.lastStart = start
        history.lastEnd = end
        ctx.emit('change', { start, end })
      }
    }

    return {
      renderProps,
      parsedValue,
      noEvents,
      move,
      next,
      prev,
      getScopeSlots,
      timeToY,
      scrollToTime,
      calendarRef,
      updateTimes,
      checkChange,
      times
    }
  },
  render() {
    const { component, props, classes } = this.renderProps

    return h(
      component,
      {
        class: ['v-calendar', classes, { 'v-calendar-events': !this.noEvents }],
        ref: el => {
          this.calendarRef = el
        },
        ...this.$props,
        ...props,
        start: props.start.date,
        end: props.end.date
      },
      this.getScopeSlots()
    )
  }
})

export default Calendar
