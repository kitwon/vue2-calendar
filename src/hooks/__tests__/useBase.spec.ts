import useBase from '../useBase'
import Calendar from '../../Calendar'
import { parseTimestamp } from '../../util/timestamp'
import { mount } from '@vue/test-utils'
import { BaseProps } from '../../util/props'
// import { ExtractVue } from '../../../../util/mixins'

export function getWrapper(props: Partial<BaseProps>) {
  return mount(Calendar, {
    props: props
  })
}

export function getProps(props: Partial<BaseProps>) {
  return getWrapper(props).vm.$props
}

describe('calendar-base.ts', () => {
  it('should parse start & end', async () => {
    const props = getProps({
      start: '2020-12-28',
      end: '2021-01-08'
    })
    const instance = useBase(props)

    expect(instance.parsedStart.value).toBeDefined()
    expect(instance.parsedStart.value).toMatchSnapshot()
    expect(instance.parsedEnd.value).toBeDefined()
    expect(instance.parsedEnd.value).toMatchSnapshot()
  })

  it('should create a day list', async () => {
    const props = getProps({
      start: '2020-12-28',
      end: '2021-01-08'
    })
    console.log(props)
    const instance = useBase(props)

    expect(instance.days.value).toBeDefined()
    expect(instance.days.value).toHaveLength(12)
    expect(instance.days.value).toMatchSnapshot()

    expect(instance.days.value[0].date).toBe('2020-12-28')
    expect(instance.days.value[10].date).toBe('2021-01-07')
  })

  it('should calculate weekday skips', async () => {
    const props = getProps({
      start: '2020-12-28',
      end: '2021-01-08'
    })
    const instance = useBase(props)

    expect(instance.weekdaySkips.value).toBeDefined()
    expect(instance.weekdaySkips.value).toHaveLength(7)
  })

  it('should generate classes', async () => {
    const props = getProps({
      start: '2020-12-28',
      end: '2021-01-08'
    })
    const instance = useBase(props)

    expect(
      instance.getRelativeClasses(parseTimestamp('2020-12-28'))
    ).toBeDefined()
    expect(
      instance.getRelativeClasses(parseTimestamp('2020-12-29'))
    ).toMatchSnapshot()
  })

  it('should generate classes with outside', async () => {
    const props = getProps({
      start: '2020-12-28',
      end: '2021-01-08'
    })
    const instance = useBase(props)

    expect(
      instance.getRelativeClasses(parseTimestamp('2020-12-28'), true)
    ).toBeDefined()
    expect(
      instance.getRelativeClasses(parseTimestamp('2020-12-28'), true)
    ).toMatchSnapshot()
  })

  it('should return weekdayFormatter equal to weekdayFormat prop', async () => {
    const weekdayFormat = (x: any) => x
    const instance = useBase({
      weekdayFormat
    })

    expect(instance.weekdayFormatter.value).toEqual(weekdayFormat)
  })

  it('should long-format weekday', async () => {
    const props = getProps({
      start: '2020-12-28',
      end: '2021-01-08'
    })
    const instance = useBase(props)

    expect(instance.weekdayFormatter.value).toBeDefined()
    expect(typeof instance.weekdayFormatter.value).toEqual('function')

    expect(
      instance.weekdayFormatter.value(parseTimestamp('2020-12-28'), false)
    ).toEqual('Monday')
    expect(
      instance.weekdayFormatter.value(parseTimestamp('2020-12-27'), false)
    ).toEqual('Sunday')
    expect(
      instance.weekdayFormatter.value(parseTimestamp('2020-12-28'), false)
    ).toEqual('Monday')
  })

  it('should short-format weekday', async () => {
    const props = getProps({})
    const instance = useBase(props)

    expect(instance.weekdayFormatter.value).toBeDefined()
    expect(typeof instance.weekdayFormatter.value).toEqual('function')

    expect(
      instance.weekdayFormatter.value(parseTimestamp('2020-12-26'), true)
    ).toEqual('Sat')
    expect(
      instance.weekdayFormatter.value(parseTimestamp('2020-12-28'), true)
    ).toEqual('Mon')
    expect(
      instance.weekdayFormatter.value(parseTimestamp('2020-12-29'), true)
    ).toEqual('Tue')
  })

  it('should get start of week', async () => {
    const props = getProps({})
    const instance = useBase(props)

    expect(
      instance.getStartOfWeek(parseTimestamp('2020-12-20')).weekday
    ).toEqual(0)
    expect(
      instance.getStartOfWeek(parseTimestamp('2019-01-27')).weekday
    ).toEqual(0)
  })

  it('should get end of week', async () => {
    const props = getProps({})
    const instance = useBase(props)

    expect(instance.getEndOfWeek(parseTimestamp('2020-12-8')).weekday).toEqual(
      6
    )
    expect(instance.getEndOfWeek(parseTimestamp('2020-12-29')).weekday).toEqual(
      6
    )
  })

  it('should return dayFormatter equal to dayFormat prop', async () => {
    const dayFormat = (x: any) => x
    const props = getProps({
      dayFormat
    })
    const instance = useBase(props)

    expect(typeof instance.dayFormatter.value).toEqual('function')
  })

  it('should format day', async () => {
    const props = getProps({
      start: '2020-12-28',
      end: '2021-01-08'
    })
    const instance = useBase(props)

    expect(instance.weekdayFormatter.value).toBeDefined()
    expect(typeof instance.weekdayFormatter.value).toEqual('function')

    expect(
      instance.dayFormatter.value(parseTimestamp('2019-01-28'), false)
    ).toEqual('28')
    expect(
      instance.dayFormatter.value(parseTimestamp('2019-01-27'), false)
    ).toEqual('27')
    expect(
      instance.dayFormatter.value(parseTimestamp('2020-12-28'), false)
    ).toEqual('28')
  })
})
