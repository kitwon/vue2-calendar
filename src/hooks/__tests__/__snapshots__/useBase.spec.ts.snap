// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendar-base.ts should create a day list 1`] = `
Array [
  Object {
    "date": "2020-12-28",
    "day": 28,
    "future": false,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 12,
    "past": true,
    "present": false,
    "time": "",
    "weekday": 1,
    "year": 2020,
  },
  Object {
    "date": "2020-12-29",
    "day": 29,
    "future": false,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 12,
    "past": false,
    "present": true,
    "time": "",
    "weekday": 2,
    "year": 2020,
  },
  Object {
    "date": "2020-12-30",
    "day": 30,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 12,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 3,
    "year": 2020,
  },
  Object {
    "date": "2020-12-31",
    "day": 31,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 12,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 4,
    "year": 2020,
  },
  Object {
    "date": "2021-01-01",
    "day": 1,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 1,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 5,
    "year": 2021,
  },
  Object {
    "date": "2021-01-02",
    "day": 2,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 1,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 6,
    "year": 2021,
  },
  Object {
    "date": "2021-01-03",
    "day": 3,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 1,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 0,
    "year": 2021,
  },
  Object {
    "date": "2021-01-04",
    "day": 4,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 1,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 1,
    "year": 2021,
  },
  Object {
    "date": "2021-01-05",
    "day": 5,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 1,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 2,
    "year": 2021,
  },
  Object {
    "date": "2021-01-06",
    "day": 6,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 1,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 3,
    "year": 2021,
  },
  Object {
    "date": "2021-01-07",
    "day": 7,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 1,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 4,
    "year": 2021,
  },
  Object {
    "date": "2021-01-08",
    "day": 8,
    "future": true,
    "hasDay": true,
    "hasTime": false,
    "hour": 0,
    "minute": 0,
    "month": 1,
    "past": false,
    "present": false,
    "time": "",
    "weekday": 5,
    "year": 2021,
  },
]
`;

exports[`calendar-base.ts should generate classes 1`] = `
Object {
  "v-future": false,
  "v-outside": false,
  "v-past": false,
  "v-present": false,
}
`;

exports[`calendar-base.ts should generate classes with outside 1`] = `
Object {
  "v-future": false,
  "v-outside": true,
  "v-past": false,
  "v-present": false,
}
`;

exports[`calendar-base.ts should parse start & end 1`] = `
Object {
  "date": "2020-12-28",
  "day": 28,
  "future": false,
  "hasDay": true,
  "hasTime": false,
  "hour": 0,
  "minute": 0,
  "month": 12,
  "past": false,
  "present": false,
  "time": "",
  "weekday": 1,
  "year": 2020,
}
`;

exports[`calendar-base.ts should parse start & end 2`] = `
Object {
  "date": "2021-01-08",
  "day": 8,
  "future": false,
  "hasDay": true,
  "hasTime": false,
  "hour": 0,
  "minute": 0,
  "month": 1,
  "past": false,
  "present": false,
  "time": "",
  "weekday": 5,
  "year": 2021,
}
`;
